### PR TITLE
1350 / Units

### PR DESCRIPTION
**Pull request by seanmalbert**
# Pull Request Template

## Issue
#1350 
Addresses # (issue)
- [ ] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

You should be able to add new and edit units for a listing. Unit Type is set to required.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Prototype/POC (not to merge)
- [ ] This change is a refactor/address technical debt
- [ ] This change requires a documentation update
- [ ] This change requires a SQL Script

## How Can This Be Tested/Reviewed?

Please describe the tests that you ran to verify your changes. Provide instructions so we can review. Please also list any relevant details for your test configuration
- [ ] Desktop View
- [ ] Mobile View
- [ ] Test A
- [ ] Test B

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have updated the changelog to include a description of my changes

### CODE REVIEW

**Comment by emilyjablonski** [comment]
* I think status should likely eventually be a dropdown. When I set the status to "Available" it shows all the units as waitlisted since the unit summaries check for (unit.status === "available”)
* The AMI Chart dropdown is showing duplicates of the TCAC 2021 chart.
* Hitting Copy and New doesn't seem to quite work. I click it once and nothing happens, and if I click it again, it works but throws an error.
* It looks like Max Occupancy is repeated in the detail view under the Eligibility header if you're viewing an individual listing.
* After I make units in a listing, save it as draft, and go back in to re-edit it, the AMI chart is not pre-filling.
* When rent is selected as % income, we should auto-set the minimum income to $0 (see below)
Screen Shot 2021-07-01 at 2 58 19 PM

File: CHANGELOG.md [hunk 1]
```diff
@@ -24,7 +24,7 @@ All notable changes to this project will be documented in this file. The format
   - Fix units availability ([#1397](https://github.com/bloom-housing/bloom/issues/1397))
 
 - Added:
-
+  - Adds units to listings ([#1350](https://github.com/bloom-housing/bloom/issues/1350))
```
Comment (emilyjablonski) [inline CHANGELOG.md:27]: The number here should be 1448 and it should probably be under the Frontend section

File: sites/partners/src/listings/PaperListingForm/index.tsx [hunk 2]
```diff
@@ -26,11 +26,14 @@ import AdditionalDetails from "./sections/AdditionalDetails"
...
 import AdditionalFees from "./sections/AdditionalFees"
+import Units from "./sections/Units"
+import { stringToBoolean, stringToNumber } from "../../../lib/helpers"
+import { useAmiChartList } from "../../../lib/hooks"
 import BuildingDetails from "./sections/BuildingDetails"
 import ListingIntro from "./sections/ListingIntro"
 import BuildingFeatures from "./sections/BuildingFeatures"
 
-type FormListing = ListingCreate & ListingUpdate
+type FormListing = Listing
```
Comment (emilyjablonski) [inline sites/partners/src/listings/PaperListingForm/index.tsx:36]: Curious why you made this change?

File: sites/partners/src/listings/PaperListingForm/index.tsx [hunk 3]
```diff
@@ -145,6 +191,45 @@ const ListingForm = ({ listing, editMode }: ListingFormProps) => {
...
+      data.disableUnitsAccordion = stringToBoolean(data.disableUnitsAccordion)
+      units.forEach((unit) => {
+        switch (unit.unitType) {
+          case "threeBdrm":
+            unit.numBedrooms = 3
+            break
+          case "twoBdrm":
+            unit.numBedrooms = 2
+            break
+          default:
```
Comment (emilyjablonski) [inline sites/partners/src/listings/PaperListingForm/index.tsx:204]: In current listing jsons, in some cases where it's a studio we are setting numBedrooms to null and in other cases we are setting it to 1, not sure which case should be settled on, but studios kinda sorta don't have bedrooms but kinda do?

File: sites/partners/src/listings/PaperListingForm/sections/Units.tsx [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+type UnitProps = {
+  units: TempUnit[]
+  setUnits: (units: TempUnit[]) => void
+  amiCharts: AmiChart[]
+  disableUnitsAccordion: boolean
+}
+
+const FormUnits = ({ units, setUnits, amiCharts, disableUnitsAccordion }: UnitProps) => {
+  const [unitDrawer, setUnitDrawer] = useState<number | null>(null)
+  // const [unitDrawerOpen, setUnitDrawerOpen] = useState<boolean>(false)
```
Comment (emilyjablonski) [inline sites/partners/src/listings/PaperListingForm/sections/Units.tsx:30]: Leftover comment maybe?

**Comment by seanmalbert** [comment]
Thank you for the feedback, @emilyjablonski !

> * I think status should likely eventually be a dropdown. When I set the status to "Available" the unit summary table on the listing page shows all the units as waitlisted since the unit summaries check for (unit.status === "available”)

I agree, and there was actually discussion in another issue that we would just leave it out, so I compromised and set it to always be "available" with a disabled field. This way we know it's there for future use.

> * The AMI Chart dropdown is showing duplicates of the TCAC 2021 chart.

Checking out the ids, they're all unique, so it's a data issue.

> * Hitting Copy and New doesn't seem to quite work. I click it once and nothing happens, and if I click it again, it works but throws an error.

Nice catch. The first time I believe was actually working for you, there's just no indication that it did, and I fixed the issue for clicking it multiple times. So now, as you click it, you should see units getting added. I think we can handle the notifications for it in a separate issue.

> * It looks like Max Occupancy is repeated in the detail view under the Eligibility header if you're viewing an individual unit.

Thanks, removed.
> * After I make units in a listing, save it as draft, and go back in to re-edit it, the AMI chart is not pre-filling.

Fixed.
> * When rent is selected as % income, we should auto-set the minimum income to $0 (see below)

Updated to set min income to 0 when % income is selected.

File: sites/partners/src/listings/PaperListingForm/sections/Units.tsx (hunk 4 above)
Comment (seanmalbert) [reply sites/partners/src/listings/PaperListingForm/sections/Units.tsx]: Yea, I removed it.

File: sites/partners/src/listings/PaperListingForm/index.tsx (hunk 3 above)
Comment (seanmalbert) [reply sites/partners/src/listings/PaperListingForm/index.tsx]: Updated per your suggestion.

File: CHANGELOG.md (hunk 1 above)
Comment (seanmalbert) [reply CHANGELOG.md]: Updated

File: sites/partners/src/listings/PaperListingForm/index.tsx (hunk 2 above)
Comment (seanmalbert) [reply sites/partners/src/listings/PaperListingForm/index.tsx:36]: I was running into issues with units and how it's currently setup with listings. I can look into this further.

**Review by emilyjablonski — APPROVED** [review]
Looks great! Thanks for the updates 👍

**Pull request merged**